### PR TITLE
feat: payment system

### PR DESCRIPTION
**Pull request by forehalo**
tracked in #4439

### CODE REVIEW

File: packages/backend/server/src/modules/payment/stripe.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import { Config } from '../../config';
+
+export const StripeProvider: FactoryProvider = {
+  provide: Stripe,
+  useFactory: (config: Config) => {
+    const stripeConfig = config.payment.stripe;
+
+    return new Stripe(
+      stripeConfig.keys.APIKey,
```
Comment (darkskygit) [inline packages/backend/server/src/modules/payment/stripe.ts:13]: does the server not consider the scene where payment is not enabled? e.g. self hosted

**Review by joooye34 — APPROVED** [review]

File: packages/backend/server/src/modules/payment/stripe.ts (hunk 1 above)
Comment (joooye34) [reply packages/backend/server/src/modules/payment/stripe.ts:13]: merge and fix later, if have some problem

**Review by darkskygit — APPROVED** [review]

File: packages/backend/server/src/modules/payment/resolver.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+      const monthly = prices.find(
+        price =>
+          decodeLookupKey(
+            // @ts-expect-error empty lookup key is filtered out
+            price.lookup_key
+          )[1] === SubscriptionRecurring.Monthly
+      );
+
+      if (!yearly || !monthly) {
+        throw new BadGatewayException(
```
Comment (Brooooooklyn) [inline packages/backend/server/src/modules/payment/resolver.ts:171]: Throw `GraphQLError` from `graphql` in GraphQL resolver: https://www.apollographql.com/docs/apollo-server/data/errors/

File: packages/backend/server/src/modules/payment/resolver.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+@Resolver(() => UserType)
+export class UserSubscriptionResolver {
+  constructor(private readonly db: PrismaService) {}
+
+  @ResolveField(() => UserSubscriptionType, { nullable: true })
+  async subscription(@CurrentUser() me: User, @Parent() user: User) {
+    if (me.id !== user.id) {
+      throw new ForbiddenException();
```
Comment (Brooooooklyn) [inline packages/backend/server/src/modules/payment/resolver.ts:243]: `GraphQLError`

File: packages/backend/server/src/modules/payment/resolver.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  @ResolveField(() => [UserInvoiceType])
+  async invoices(
+    @CurrentUser() me: User,
+    @Parent() user: User,
+    @Args('take', { type: () => Int, nullable: true, defaultValue: 8 })
+    take: number,
+    @Args('skip', { type: () => Int, nullable: true }) skip?: number
+  ) {
+    if (me.id !== user.id) {
+      throw new ForbiddenException();
```
Comment (Brooooooklyn) [inline packages/backend/server/src/modules/payment/resolver.ts:262]: `GraphQLError`

File: packages/backend/server/src/modules/payment/schedule.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+  get isActive() {
+    return this._schedule?.status === 'active';
+  }
+
+  async fromSchedule(schedule: string | Stripe.SubscriptionSchedule) {
+    if (typeof schedule === 'string') {
+      const s = await this.stripe.subscriptionSchedules
+        .retrieve(schedule)
+        .catch(() => undefined);
```
Comment (Brooooooklyn) [inline packages/backend/server/src/modules/payment/schedule.ts:53]: should we add log here?

File: packages/backend/server/src/modules/payment/service.ts [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+        user: true,
+      },
+    });
+
+    if (customer?.user) {
+      return customer.user;
+    }
+
+    // customer may not saved is db, check it with stripe
+    const stripeCustomer = await this.stripe.customers.retrieve(customerId);
```
Comment (darkskygit) [inline packages/backend/server/src/modules/payment/service.ts:581]: idempotent key

**Review by joooye34 — DISMISSED** [review]

File: packages/backend/server/src/modules/payment/service.ts [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+        userId: id,
+      },
+    });
+
+    if (!user) {
+      throw new Error('Unknown user');
+    }
+
+    try {
+      const portal = await this.stripe.billingPortal.sessions.create({
```
[the remark this answers is not in the document]
Comment (joooye34) [reply packages/backend/server/src/modules/payment/service.ts:311]: improve later

File: packages/backend/server/src/modules/payment/service.ts [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+      price,
+      // if user is early access user, use early access coupon
+      manager.currentPhase?.coupon === CouponType.EarlyAccess ||
+        manager.currentPhase?.coupon === CouponType.EarlyAccessRenew ||
+        manager.nextPhase?.coupon === CouponType.EarlyAccessRenew
+        ? CouponType.EarlyAccessRenew
+        : undefined
+    );
+
+    return await this.db.userSubscription.update({
```
[the remark this answers is not in the document]
Comment (joooye34) [reply packages/backend/server/src/modules/payment/service.ts:288]: improve later

**Review by darkskygit — APPROVED** [review]

**Review by joooye34 — APPROVED** [review]

**Review by darkskygit — APPROVED** [review]

**Review by joooye34 — APPROVED** [review]

**Pull request merged**